### ISSUE TITLE
feature wish: multiple input devices

### ISSUE THREAD

**Issue by tiborgats** [redox-os/redox#636]
I wish for the possibility to be able to handle two mouses, two keyboards, or more. For example if someone already has two monitors, it would be nice for two users to be able to work simultaneously on the same machine by connecting an extra mouse & keyboard. (Even if one of them is working inside a virtual machine.)
It would also open new possibilities for gaming if it were possible to use two mouses, for both hands at the same time.

**Comment by ca1ek** [comment]
You mean something like, one keyboard writing in one window i.e. Sodium editor while the other keyboard is writing in another window i.e. the terminal?

**Comment by tormeh** [comment]
Would also be cool for split screen multiplayer!

**Comment by tiborgats** [comment]
I would consider many use case possibilities. However, I'm not sure how difficult (or possible) is to implement them. I thought of the followings:
1. One user is logged in, but two keyboards and mouses are assigned to two different applications. This case would allow the split-screen multiplayer games. Or in a dual-screen setup someone can work in a virtual machine on one screen while the other is working on the other screen on other applications.
2. One user logged in, but uses two mouses (uses both hands) to control the same application. E.g. in a FPS game, he can control the movement of the character with one mouse while the other mouse controls the aiming of his weapon. This would also allow two (or more) people cooperating on a [synthesizer](http://zynaddsubfx.sourceforge.net/), with multiple mouses.
3. Two users are logged in at the same time (splitting up hardware resources), one works on one screen while the other one on the other screen. E.g. my computer has dual screen, normally I use both of them, but if my wife wants to write a document while I am working I would let her use one of the screens, so we could work at the same time (if it were possible). For these kind of situations having two computers instead of just using an extra set of mouse+keyboard is really a waste of money.

**Comment by qm3ster** [comment]
"Focus" needs to be considered. I suggest the following for the mentioned cases:

1. Groups of input devices can be assigned to represent "Humans" In this case. That way each gets an active window, a focused control, a pointer, and a text cursor. There is still the issue of window layering, and the **usefulness of this approach seem very dubious** at current display configurations.
- Most people are incapable of operating two mice at the same time, but this can be supported through the native multitouch module (although current touch screens usually don't result in 'multi-hover' like this approach would.) and on a per-application basis, either through multitouch, or, considering the big difference in usage, through tracking the events of a specific mouse.
- This is the sanest approach and requires the least amount of modification that would later give developers headaches - two desktop environments are run, each displaying to its own screen and filtering standard input devices or USB ports.
 - This could be a considerable performance improvement over running multiple virtual machines, and would allow the sharing of a display adapter without any extra tricks. 
 - It does necessitate concurrent active user sessions though for reasons such as security and singleton application sessions (browsers/IDEs/etc.)
   - Fortunately, these are inevitable on a desktop machine, especially with a microkernel, since you will need users for running services and maintenance processes that are protected from each other.

**Comment by rushsteve1** [comment]
Overall I doubt the usefulness of this outside of a few fringe cases.
However in those cases the best way to do it would probably be having multiple users logged in, each with their own devices/displays. This would be more hardware intensive, but is probably the most effective solution.
However I do like @qm3ster's idea of multitouch. This would perhaps allow for multiple mice as stated, as well as allowing a good foundation for touchscreens.

**Comment by jawbonegroove** [comment]
I don't think it would be all that difficult to assign extra keyboards to windows or displays, and honestly this would simplify my life a lot.  and multitouch support will offer a basis for allowing multiple mice, for sure.

even

**Comment by rushsteve1** [comment]
The larger issue here, as I see it, would be having multiple windows in focus at the same time and maintaining which inputs are resolved to what window.

**Comment by qm3ster** [comment]
@rushsteve1 that's why this requires a collection of "Humans", each containing _an active window, a focused control, a pointer, and a text cursor_. It doesn't seem like it would introduce any serious de-optimizations, layout would still be a single state per screen. Not sure who would need this, still.

**Comment by Mouvedia** [comment]
The buzz factor of a youtube video demonstrating the possibilities would be tremendous for the project.

**Comment by rushsteve1** [comment]
@qm3ster I get what you're saying now. Why not tie this "humans" idea to logins. Users could login to an existing session and collaborate (this would also have reduced overhead because we wouldn't need 2 Orbital sessions, etc.) or into a new session, allowing access to shared resources while also being separated. (https://en.m.wikipedia.org/wiki/Multiseat_configuration)

If we were to have a system of sandboxing applications we could have various things such as mouse and keyboard data passed to applications based on who is controlling them.

An issue however would be when multiple users try to access the same file at the same time.

@Mouvedia the popularity would be nice but I think this is still a ways off :smile:

**Comment by qm3ster** [comment]
Whose version of applications (overwritten user configs) are you going to run?
What happens to singleton applications like a text editor or a browser that have per user persistent tabs and open new URLs alongside them?
The whole idea is allowing extra pairs of hands into the same user.
The only time that crazy idea would be useful is if you want to move windows from one device to a whole other device, for example to someone else's computer. That is a basically an impossible security challenge in a trustless environment. In a trusted environment, this is much easier achieved by applications that pass compact business state to a different version of the application on the different device, and so depends on developers of end applications, not the platform.

**Comment by Sherif89** [comment]
I think [multiseat configuration (Or multiterminal)](https://en.wikipedia.org/wiki/Multiseat_configuration) is better. It is a single computer which supports multiple independent local users at the same time.

**Comment by Sherif89** [comment]
Sorry, I wasn't aware that the issue was left for a long time, before commenting.
